storage/txn_cache.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2014 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: yananzhi (zac.zhiyanan@gmail.com)
+
+package storage
+
+import (
+	"sync"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util/cache"
+
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+const (
+	// The default size of the txn cache.
+	defaultTxnCacheSize = 1 << 20
+)
+
+// TxnCache maintains txn which was pushed recently to avoid
+// repushing further intents after a txn has already been aborted,
+// comiitted or its timestamp moved forward.
+type TxnCache struct {
+	// Use a LRU un-ordered cache, key is txn's ID, value is txn
+	txnCache *cache.UnorderedCache
+
+	// cacheMu protects txnCache for concurrent access
+	cacheMu sync.Mutex
+}
+
+// NewTxnCache returns a new txn cache
+func NewTxnCache(maxSize int) *TxnCache {
+
+	return &TxnCache{
+		txnCache: cache.NewUnorderedCache(cache.Config{
+			Policy: cache.CacheLRU,
+			ShouldEvict: func(n int, k, v interface{}) bool {
+				return n > maxSize
+			},
+		}),
+	}
+}
+
+// processPushedTxn return true if the cache txn can process a push txn request, and vice versa
+func (tc *TxnCache) processPushedTxn(args proto.Request, reply proto.Response) bool {
+
+	pushTxnReq, ok := args.(*proto.InternalPushTxnRequest)
+	if !ok {
+		return false
+	}
+
+	tc.cacheMu.Lock()
+	defer tc.cacheMu.Unlock()
+
+	value, ok := tc.txnCache.Get(string(pushTxnReq.PusheeTxn.ID))
+
+	if !ok {
+		return false
+	}
+
+	// Compare cached txn and pushee txn
+	cachedTxn := value.(*proto.Transaction)
+	if sufficientPushTxnRequest(cachedTxn, &pushTxnReq.PusheeTxn, pushTxnReq.PushType) {
+		pushTxnReply := reply.(*proto.InternalPushTxnResponse)
+
+		pushTxnReply.PusheeTxn = gogoproto.Clone(cachedTxn).(*proto.Transaction)
+
+		// Upgrade the epoch and priority as necessary.
+		if pushTxnReply.PusheeTxn.Epoch < pushTxnReq.PusheeTxn.Epoch {
+			pushTxnReply.PusheeTxn.Epoch = pushTxnReq.PusheeTxn.Epoch
+			// Also upgrade the cached txn's epoch
+			cachedTxn.Epoch = pushTxnReply.PusheeTxn.Epoch
```

Comment:
@bdarnell  Txn's epoch and priority is also keeped in txn current reuqest, cached txn lost this information seems has no side effect.